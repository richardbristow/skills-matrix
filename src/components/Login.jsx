import React, { useState } from 'react';
import {
  Button,
  Form,
  Card,
  Alert,
  InputGroup,
  Spinner,
} from 'react-bootstrap';
import PropTypes from 'prop-types';
import { Auth } from 'aws-amplify';
import { AtSign, Lock } from 'react-feather';
import styled from 'styled-components/macro';

import StyledMain from '../shared/StyledMain';

const StyledLogin = styled(StyledMain)`
  @media all and (min-width: 480px) {
    max-width: 400px;
  }
`;

const Login = ({ setAuthenticated, ...props }) => {
  const [values, setValues] = useState({ email: '', password: '' });
  const [validated, setValidated] = useState(false);
  const [authError, setAuthError] = useState(null);
  const [loading, setLoading] = useState(false);

  const { email, password } = values;

  const handleInputChange = ({ target }) => {
    const { name, value } = target;
    setValues({ ...values, [name]: value });
    setAuthError(null);
  };

  const handleSubmit = async event => {
    event.preventDefault();
    setAuthError(null);
    setLoading(true);
    const form = event.currentTarget;

    if (form.checkValidity() === false) {
      event.stopPropagation();
      setLoading(false);
      setValidated(true);
    } else {
      try {
        await Auth.signIn(email, password);
        setAuthenticated(true);
        props.history.push('/');
      } catch (err) {
        setAuthError(err.message);
        setLoading(false);
        setValidated(true);
      }
    }
  };

  const checkForEmptyForm = () => {
    return !(email.length > 0 && password.length > 0);
  };

  return (
    <StyledLogin centre>
      <Card>
        <Card.Header as="h3">Login</Card.Header>
        <Card.Body>
          <Form validated={validated} noValidate onSubmit={handleSubmit}>
            <Form.Group controlId="email">
              <Form.Label>Email address</Form.Label>
              <InputGroup>
                <InputGroup.Prepend>
                  <InputGroup.Text>
                    <AtSign size={16} />
                  </InputGroup.Text>
                </InputGroup.Prepend>
                <Form.Control
                  value={email}
                  autoFocus
                  onChange={handleInputChange}
                  name="email"
                  type="email"
                  placeholder="Enter email..."
                  required
                />
                <Form.Control.Feedback type="invalid">
                  Please provide a valid email address.
                </Form.Control.Feedback>
              </InputGroup>
            </Form.Group>
            <Form.Group controlId="password">
              <Form.Label>Password</Form.Label>
              <InputGroup>
                <InputGroup.Prepend>
                  <InputGroup.Text>
                    <Lock size={16} />
                  </InputGroup.Text>
                </InputGroup.Prepend>
                <Form.Control
                  value={password}
                  onChange={handleInputChange}
                  name="password"
                  type="password"
                  placeholder="Password..."
                  required
                />
                <Form.Control.Feedback type="invalid">
                  Please provide a password.
                </Form.Control.Feedback>
              </InputGroup>
            </Form.Group>
            {authError && (
              <Alert variant="danger">
                <strong>Error:</strong> {authError}
              </Alert>
            )}
            <Button
              variant="outline-primary"
              disabled={checkForEmptyForm() || loading}
              type="submit"
              block
            >
              {loading ? (
                <>
                  <Spinner
                    as="span"
                    animation="grow"
                    variant="primary"
                    size="sm"
                    role="status"
                    aria-hidden="true"
                  />
                  Logging in...
                </>
              ) : (
                'Login'
              )}
            </Button>
          </Form>
        </Card.Body>
      </Card>
    </StyledLogin>
  );
};

Login.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  history: PropTypes.object.isRequired,
  setAuthenticated: PropTypes.func.isRequired,
};

export default Login;
