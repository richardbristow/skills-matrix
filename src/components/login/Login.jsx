import React, { useState } from 'react';
import { Card } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { Auth } from 'aws-amplify';
import styled from 'styled-components/macro';

import StyledMain from '../../shared/StyledMain';
import ChangePasswordForm from './ChangePasswordForm';
import LoginForm from './LoginForm';

const StyledLogin = styled(StyledMain)`
  @media all and (min-width: 480px) {
    max-width: 400px;
  }
`;

const Login = ({ setAuthenticated, setAuthenticating, ...props }) => {
  const [values, setValues] = useState({
    email: '',
    password: '',
    newPassword: '',
    confirmPassword: '',
  });
  const [validated, setValidated] = useState(false);
  const [isError, setIsError] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [changePassword, setChangePassword] = useState(false);

  const { email, password, newPassword, confirmPassword } = values;

  const handleInputChange = ({ target }) => {
    const { name, value } = target;
    setValues({ ...values, [name]: value });
    setIsError(null);
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    setIsError(null);
    setIsLoading(true);
    const form = event.currentTarget;

    if (form.checkValidity() === false) {
      event.stopPropagation();
      setIsLoading(false);
      setValidated(true);
    } else {
      try {
        const user = await Auth.signIn(email, password);
        if (user.challengeName === 'NEW_PASSWORD_REQUIRED') {
          if (changePassword) {
            if (newPassword !== confirmPassword) {
              setIsError({ message: 'Passwords must match.' });
              setIsLoading(false);
              setValidated(true);
            } else {
              await Auth.completeNewPassword(user, newPassword);
              setChangePassword(false);
              setAuthenticated(true);
              setAuthenticating(true);
              props.history.push('/');
            }
          } else {
            setChangePassword(true);
            setIsLoading(false);
          }
        } else {
          setAuthenticated(true);
          setAuthenticating(true);
          props.history.push('/');
        }
      } catch (err) {
        setIsError(err);
        setIsLoading(false);
        setValidated(true);
      }
    }
  };

  const checkForEmptyForm = () => {
    return !(email.length > 0 && password.length > 0);
  };

  return (
    <StyledLogin centre>
      <Card>
        <Card.Header as="h3">
          {changePassword ? 'Change Password' : 'Login'}
        </Card.Header>
        {changePassword ? (
          <ChangePasswordForm
            validated={validated}
            handleSubmit={handleSubmit}
            handleInputChange={handleInputChange}
            checkForEmptyForm={checkForEmptyForm}
            isError={isError}
            isLoading={isLoading}
            values={values}
          />
        ) : (
          <LoginForm
            validated={validated}
            handleSubmit={handleSubmit}
            handleInputChange={handleInputChange}
            checkForEmptyForm={checkForEmptyForm}
            isError={isError}
            isLoading={isLoading}
            values={values}
          />
        )}
      </Card>
    </StyledLogin>
  );
};

Login.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  history: PropTypes.object.isRequired,
  setAuthenticated: PropTypes.func.isRequired,
  setAuthenticating: PropTypes.func.isRequired,
};

export default Login;
