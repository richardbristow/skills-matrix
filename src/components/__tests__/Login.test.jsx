import React from 'react';
import { shallow } from 'enzyme';
import Login from '../Login';

describe('Login', () => {
  it('should render the login form without throwing an error', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an email input', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find({ controlId: 'email' })).toHaveLength(1);
  });

  it('should render a password input', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find({ controlId: 'password' })).toHaveLength(1);
  });

  it('should render a login/submit button', () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find({ type: 'submit' })).toHaveLength(1);
  });
});

describe('checkForEmptyForm', () => {
  it('should disable the login button if email and password fields are empty', () => {
    const wrapper = shallow(<Login />);
    const button = wrapper.find({ type: 'submit' });
    expect(wrapper.instance().checkForEmptyForm()).toEqual(true);
    expect(button.props().disabled).toEqual(true);
  });

  it('should disable the login button if email field is empty', () => {
    const wrapper = shallow(<Login />);
    wrapper.setState({ email: 'emailAddress' });
    const button = wrapper.find({ type: 'submit' });
    expect(wrapper.instance().checkForEmptyForm()).toEqual(true);
    expect(button.props().disabled).toEqual(true);
  });

  it('should disable the login button if password field is empty', () => {
    const wrapper = shallow(<Login />);
    wrapper.setState({ password: 'password' });
    const button = wrapper.find({ type: 'submit' });
    expect(wrapper.instance().checkForEmptyForm()).toEqual(true);
    expect(button.props().disabled).toEqual(true);
  });

  it('should enable the login button if email and password fields have values', () => {
    const wrapper = shallow(<Login />);
    wrapper.setState({ email: 'emailAddress', password: 'password' });
    const button = wrapper.find({ type: 'submit' });
    const emailInput = wrapper.find({ type: 'email' });
    const passwordInput = wrapper.find({ type: 'password' });
    expect(emailInput.props().value).toEqual('emailAddress');
    expect(passwordInput.props().value).toEqual('password');
    expect(wrapper.instance().checkForEmptyForm()).toEqual(false);
    expect(button.props().disabled).toEqual(false);
  });
});
