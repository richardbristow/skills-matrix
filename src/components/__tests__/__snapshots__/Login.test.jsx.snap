// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the login form without throwing an error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h2>
          Login
        </h2>,
        <ForwardRef(Bootstrap(Form))
          className="justify-content-center"
          noValidate={true}
          onSubmit={[Function]}
          validated={false}
        >
          <ForwardRef(Bootstrap(FormGroup))
            controlId="email"
          >
            <ForwardRef(ContextTransform)
              autoFocus={true}
              onChange={[Function]}
              placeholder="Email"
              required={true}
              type="email"
              value=""
            />
            <Feedback
              as="div"
              type="invalid"
            >
              Please provide a valid email address.
            </Feedback>
          </ForwardRef(Bootstrap(FormGroup))>
          <ForwardRef(Bootstrap(FormGroup))
            controlId="password"
          >
            <ForwardRef(ContextTransform)
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
              value=""
            />
            <Feedback
              as="div"
              type="invalid"
            >
              Please provide a password.
            </Feedback>
          </ForwardRef(Bootstrap(FormGroup))>
          <ForwardRef(Bootstrap(Button))
            block={true}
            disabled={true}
            type="submit"
            variant="outline-primary"
          >
            Login
          </ForwardRef(Bootstrap(Button))>
        </ForwardRef(Bootstrap(Form))>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
        },
        "ref": null,
        "rendered": "Login",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ForwardRef(Bootstrap(FormGroup))
              controlId="email"
            >
              <ForwardRef(ContextTransform)
                autoFocus={true}
                onChange={[Function]}
                placeholder="Email"
                required={true}
                type="email"
                value=""
              />
              <Feedback
                as="div"
                type="invalid"
              >
                Please provide a valid email address.
              </Feedback>
            </ForwardRef(Bootstrap(FormGroup))>,
            <ForwardRef(Bootstrap(FormGroup))
              controlId="password"
            >
              <ForwardRef(ContextTransform)
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
                value=""
              />
              <Feedback
                as="div"
                type="invalid"
              >
                Please provide a password.
              </Feedback>
            </ForwardRef(Bootstrap(FormGroup))>,
            <ForwardRef(Bootstrap(Button))
              block={true}
              disabled={true}
              type="submit"
              variant="outline-primary"
            >
              Login
            </ForwardRef(Bootstrap(Button))>,
          ],
          "className": "justify-content-center",
          "noValidate": true,
          "onSubmit": [Function],
          "validated": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <ForwardRef(ContextTransform)
                  autoFocus={true}
                  onChange={[Function]}
                  placeholder="Email"
                  required={true}
                  type="email"
                  value=""
                />,
                <Feedback
                  as="div"
                  type="invalid"
                >
                  Please provide a valid email address.
                </Feedback>,
              ],
              "controlId": "email",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "autoFocus": true,
                  "onChange": [Function],
                  "placeholder": "Email",
                  "required": true,
                  "type": "email",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "Feedback": [Function],
                  "defaultProps": undefined,
                  "propTypes": undefined,
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "div",
                  "children": "Please provide a valid email address.",
                  "type": "invalid",
                },
                "ref": null,
                "rendered": "Please provide a valid email address.",
                "type": [Function],
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <ForwardRef(ContextTransform)
                  onChange={[Function]}
                  placeholder="Password"
                  required={true}
                  type="password"
                  value=""
                />,
                <Feedback
                  as="div"
                  type="invalid"
                >
                  Please provide a password.
                </Feedback>,
              ],
              "controlId": "password",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Password",
                  "required": true,
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "Feedback": [Function],
                  "defaultProps": undefined,
                  "propTypes": undefined,
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "div",
                  "children": "Please provide a password.",
                  "type": "invalid",
                },
                "ref": null,
                "rendered": "Please provide a password.",
                "type": [Function],
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "block": true,
              "children": "Login",
              "disabled": true,
              "type": "submit",
              "variant": "outline-primary",
            },
            "ref": null,
            "rendered": "Login",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "Check": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Input": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "Label": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
          "Control": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Feedback": [Function],
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
          "Group": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
          "Label": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
          "Row": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
          "Text": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
          "defaultProps": undefined,
          "propTypes": undefined,
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Login__StyledLogin-sc-4kj774-0",
        "isStatic": true,
        "rules": Array [
          "@media all and (min-width:480px){max-width:350px;padding-top:80px;margin:0 auto;}",
        ],
      },
      "displayName": "Login__StyledLogin",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "Login__StyledLogin-sc-4kj774-0",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h2>
            Login
          </h2>,
          <ForwardRef(Bootstrap(Form))
            className="justify-content-center"
            noValidate={true}
            onSubmit={[Function]}
            validated={false}
          >
            <ForwardRef(Bootstrap(FormGroup))
              controlId="email"
            >
              <ForwardRef(ContextTransform)
                autoFocus={true}
                onChange={[Function]}
                placeholder="Email"
                required={true}
                type="email"
                value=""
              />
              <Feedback
                as="div"
                type="invalid"
              >
                Please provide a valid email address.
              </Feedback>
            </ForwardRef(Bootstrap(FormGroup))>
            <ForwardRef(Bootstrap(FormGroup))
              controlId="password"
            >
              <ForwardRef(ContextTransform)
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
                value=""
              />
              <Feedback
                as="div"
                type="invalid"
              >
                Please provide a password.
              </Feedback>
            </ForwardRef(Bootstrap(FormGroup))>
            <ForwardRef(Bootstrap(Button))
              block={true}
              disabled={true}
              type="submit"
              variant="outline-primary"
            >
              Login
            </ForwardRef(Bootstrap(Button))>
          </ForwardRef(Bootstrap(Form))>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef(Bootstrap(FormGroup))
                controlId="email"
              >
                <ForwardRef(ContextTransform)
                  autoFocus={true}
                  onChange={[Function]}
                  placeholder="Email"
                  required={true}
                  type="email"
                  value=""
                />
                <Feedback
                  as="div"
                  type="invalid"
                >
                  Please provide a valid email address.
                </Feedback>
              </ForwardRef(Bootstrap(FormGroup))>,
              <ForwardRef(Bootstrap(FormGroup))
                controlId="password"
              >
                <ForwardRef(ContextTransform)
                  onChange={[Function]}
                  placeholder="Password"
                  required={true}
                  type="password"
                  value=""
                />
                <Feedback
                  as="div"
                  type="invalid"
                >
                  Please provide a password.
                </Feedback>
              </ForwardRef(Bootstrap(FormGroup))>,
              <ForwardRef(Bootstrap(Button))
                block={true}
                disabled={true}
                type="submit"
                variant="outline-primary"
              >
                Login
              </ForwardRef(Bootstrap(Button))>,
            ],
            "className": "justify-content-center",
            "noValidate": true,
            "onSubmit": [Function],
            "validated": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <ForwardRef(ContextTransform)
                    autoFocus={true}
                    onChange={[Function]}
                    placeholder="Email"
                    required={true}
                    type="email"
                    value=""
                  />,
                  <Feedback
                    as="div"
                    type="invalid"
                  >
                    Please provide a valid email address.
                  </Feedback>,
                ],
                "controlId": "email",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "autoFocus": true,
                    "onChange": [Function],
                    "placeholder": "Email",
                    "required": true,
                    "type": "email",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "Feedback": [Function],
                    "defaultProps": undefined,
                    "propTypes": undefined,
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "div",
                    "children": "Please provide a valid email address.",
                    "type": "invalid",
                  },
                  "ref": null,
                  "rendered": "Please provide a valid email address.",
                  "type": [Function],
                },
              ],
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "propTypes": undefined,
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <ForwardRef(ContextTransform)
                    onChange={[Function]}
                    placeholder="Password"
                    required={true}
                    type="password"
                    value=""
                  />,
                  <Feedback
                    as="div"
                    type="invalid"
                  >
                    Please provide a password.
                  </Feedback>,
                ],
                "controlId": "password",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Password",
                    "required": true,
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "Feedback": [Function],
                    "defaultProps": undefined,
                    "propTypes": undefined,
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "div",
                    "children": "Please provide a password.",
                    "type": "invalid",
                  },
                  "ref": null,
                  "rendered": "Please provide a password.",
                  "type": [Function],
                },
              ],
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "propTypes": undefined,
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "block": true,
                "children": "Login",
                "disabled": true,
                "type": "submit",
                "variant": "outline-primary",
              },
              "ref": null,
              "rendered": "Login",
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "propTypes": undefined,
                "render": [Function],
              },
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Check": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Input": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "propTypes": undefined,
                "render": [Function],
              },
              "Label": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "propTypes": undefined,
                "render": [Function],
              },
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "Control": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Feedback": [Function],
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "Group": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "Label": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "Row": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "Text": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "propTypes": undefined,
              "render": [Function],
            },
            "defaultProps": undefined,
            "propTypes": undefined,
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Login__StyledLogin-sc-4kj774-0",
          "isStatic": true,
          "rules": Array [
            "@media all and (min-width:480px){max-width:350px;padding-top:80px;margin:0 auto;}",
          ],
        },
        "displayName": "Login__StyledLogin",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Login__StyledLogin-sc-4kj774-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
